File: pkg/dsp/dft.go
package dsp

import (
	"math"
)

// DFT returns the spectrum of the signal. Note that the signal length must be 2^N, e.g. 1024.
func DFT(input []complex128) []complex128 {
	output := make([]complex128, len(input))

	for i := 0; i < len(input); i++ {
		var realSum float64
		var imagSum float64

		for j := 0; j < len(input); j++ {
			theta := 2.0 * math.Pi / float64(len(input)) * float64(i) * float64(j)

			realSum += real(input[j]) * math.Cos(theta)
			imagSum -= real(input[j]) * math.Sin(theta)
		}

		output[i] = complex(realSum, imagSum)
	}

	return output
}

File: pkg/dsp/testdata/dft.cpp
#include <complex>
#include <vector>
#include <iostream>

using namespace std;

void DFT1(
  const vector<double> a,
  vector<double>& real,
  vector<double>& imag
) {
  int n = a.size();

  for (int i = 0; i < n; i++) {
    double realSum = 0.0;
    double imagSum = 0.0;
    for (int j = 0; j < n; j++) {
      double theta = 2.0 * M_PI / static_cast<double>(n) * static_cast<double>(i) * static_cast<double>(j);
      realSum += a[j] * cos(theta);
      imagSum -= a[j] * sin(theta);
    }
    real.push_back(realSum);
    imag.push_back(imagSum);
  }
  return;
}

void DFT2(
  const vector<double> a,
  vector< complex<double> >& output
) {
  int n = a.size();

  for (int i = 0; i < n; i++) {
    complex<double> sum(0.0, 0.0);

    for (int j = 0; j < n; j++) {
      double theta = 2.0 * M_PI / static_cast<double>(n) * static_cast<double>(i) * static_cast<double>(j);
      sum += a[j] * exp(complex<double>(0.0, -theta));
    }
    output.push_back(sum);
  }
  return;
}

int main() {
  vector<double> a;
  vector<double> real;
  vector<double> imag;

  for (int i = 0; i < 32; i++) {
    a.push_back(static_cast<double>(i));
  }

  DFT1(a, real, imag);
  for (int i = 0; i < 32; i++) {
    // cout << real[i] << "\t" << imag[i] << endl;
    printf("%.3f\t%.3f\n", real[i], imag[i]);
  }
  return 0;
}
